### ISSUE TITLE
Extending color choice

### ISSUE THREAD

**Issue by salve-amice** [ACCLAB/dabestr#55]
Hello,

The issue: I needed flexibility in the choice of colours for the raw data in an estimation plot AND a colour vector long enough to cover a number of data groups >11, the number of colours in the bigger RColorBrewer palettes. The following worked for me. In code file R/plot.R from GitHub, I replaced the original line 452,

`      ggplot2::scale_color_brewer(palette = palette) +
`

with

`    ggplot2::scale_colour_manual(values = rep(c("#70A4CE", "#70A4CE", "#CF955C", "#CF955C"), 3)) +
`

This was suitable for 12 data groups paired by color, alternating blue and dark orange as in the DABEST paper by Ho et al., Nature Methods, 2019. I'm running dabestr_0.2.2 from CRAN, and had to read in the (non-modified) plot_helpers.R and flat_violin.R files to make this work, in addition to the edited plot.R file.

**Comment by josesho** [comment]
OK, thanks for pointing our where we need to modify this. This issue (greater freedom in specifying theme and palette has been on the radar (See #33 and #47). If you have a fix ready feel free to submit a PR!

**Comment by salve-amice** [comment]
You're very welcome. I must admit I'm on a learning curve as 
regards writing R functions and GitHub but I'll give the code 
for colour a go as soon as I have the time.

The current version of dabestr admits themes from packages other 
than the default ggplot2 without modification, and I found 
plenty to try out in package ggthemes. I have a weak spot for 
the simplicity of ggthemes::theme_tufte().

Best wishes,

Jose


On Monday, September 16, 2019 at 19:18 (GMT-0700), Joses W. Ho wrote:

|  OK, thanks for pointing our where we need to modify this. This
|  issue (greater freedom in specifying theme and palette has
|  been on the radar (See #33 and #47). If you have a fix ready
|  feel free to submit a PR!
|


-- 
José Enrique Mejía, PhD
Inserm U1043
CPTP, Bâtiment B, Hôpital Purpan
BP 3028, 31024 Toulouse Cedex 3
France

Tel. +33 562 744 555
Fax  +33 562 744 558

**Comment by lczubiri** [comment]
Hi all,

I have more or less the same issue as salve-amice: I needed to cover 15 data groups, existing palettes didn't work for my case (only <11 data groups) and I tried creating my own palette and then using it in plot(e.g., multi.group,palette="myownpalette") and also "scale_fill_manual()" as in ggplots. Nothing worked. I also tried what salva-amice suggests above (preplacing original line 452...) but that didn't work for me either (or I couldn't be able to make it work).

Below is an example of my data structure (5 species, 3 treatments) and the functions I'm trying to use. Any help would be very much appreciated. Thanks in advance! Cheers, Laura
library(dabestr)
#> Loading required package: boot
#> Loading required package: magrittr

sampledata<-read.table("~/sample_data.txt",  header= T, sep = "\t",dec=".", na.strings="NA", row.names=1)
sampledata
#>     Group var1
#> 1   Sp1_C    5
#> 2   Sp1_C    7
#> 3   Sp1_C    1
#> 4   Sp1_C    9
#> 5   Sp1_C    3
#> 6   Sp2_C    4
#> 7   Sp2_C    1
#> 8   Sp2_C    6
#> 9   Sp2_C    2
#> 10  Sp2_C    8
#> 11  Sp3_C    1
#> 12  Sp3_C    2
#> 13  Sp3_C   10
#> 14  Sp3_C    1
#> 15  Sp3_C    8
#> 16  Sp4_C    4
#> 17  Sp4_C    1
#> 18  Sp4_C    3
#> 19  Sp4_C    7
#> 20  Sp4_C    4
#> 21  Sp5_C    4
#> 22  Sp5_C    4
#> 23  Sp5_C    4
#> 24  Sp5_C    5
#> 25  Sp5_C    1
#> 26 Sp1_T1    7
#> 27 Sp1_T1    7
#> 28 Sp1_T1    6
#> 29 Sp1_T1    4
#> 30 Sp1_T1    9
#> 31 Sp2_T1    2
#> 32 Sp2_T1    1
#> 33 Sp2_T1    6
#> 34 Sp2_T1    8
#> 35 Sp2_T1    6
#> 36 Sp3_T1    6
#> 37 Sp3_T1    5
#> 38 Sp3_T1    3
#> 39 Sp3_T1    3
#> 40 Sp3_T1    1
#> 41 Sp4_T1    8
#> 42 Sp4_T1    5
#> 43 Sp4_T1    7
#> 44 Sp4_T1    3
#> 45 Sp4_T1    4
#> 46 Sp5_T1    6
#> 47 Sp5_T1    5
#> 48 Sp5_T1    5
#> 49 Sp5_T1   10
#> 50 Sp5_T1    7
#> 51 Sp1_T2    8
#> 52 Sp1_T2    9
#> 53 Sp1_T2    8
#> 54 Sp1_T2    4
#> 55 Sp1_T2    4
#> 56 Sp2_T2    1
#> 57 Sp2_T2    5
#> 58 Sp2_T2    7
#> 59 Sp2_T2    4
#> 60 Sp2_T2    4
#> 61 Sp3_T2    6
#> 62 Sp3_T2    3
#> 63 Sp3_T2    4
#> 64 Sp3_T2   10
#> 65 Sp3_T2    6
#> 66 Sp4_T2    7
#> 67 Sp4_T2    8
#> 68 Sp4_T2    4
#> 69 Sp4_T2    6
#> 70 Sp4_T2    3
#> 71 Sp5_T2    4
#> 72 Sp5_T2    8
#> 73 Sp5_T2    5
#> 74 Sp5_T2    3
#> 75 Sp5_T2    5

multi.group <- 
  sampledata %>%
  dabest(Group, var1, 
         idx = list(c("Sp1_C", "Sp2_C","Sp3_C","Sp4_C","Sp5_C"),
                    c("Sp1_T1", "Sp2_T1","Sp3_T1","Sp4_T1","Sp5_T1"),
                    c("Sp1_T2", "Sp2_T2","Sp3_T2","Sp4_T2","Sp5_T2")), 
         paired = FALSE)
multi.group

**Comment by salve-amice** [comment]
lczubiri, I think there are two different issues here. First, dabest() syntax. If you want to compare in each instance [species X, treatment Y] minus [species X, control] then you need to build your idx list like this:

`idx = list(`
`    c("Sp1_C", "Sp1_T1", "Sp1_T2"),`
`    c("Sp2_C", "Sp2_T1", "Sp2_T2"),`
`    c("Sp3_C", "Sp3_T1", "Sp3_T2"),`
`    c("Sp4_C", "Sp4_T1", "Sp4_T2"),`
`    c("Sp5_C", "Sp5_T1", "Sp5_T2")`
`)`

For colour specification, I saved to the working directory the three GitHub files: plot.R", flat_violin.R, and plot_helpers.R, then I edited line 452 of plot.R for a token palette from package viridis:

`viridis::scale_color_viridis(option = "inferno", discrete = TRUE, begin = 0, end = 0.9) +`

And then to plot the analysis to pdf:

`source("plot.R")`
`source("flat_violin.R")`
`source("plot_helpers.R")`
`pdf(width = 11, height = 4.5)`
`plot(multi.group)`
`dev.off()`

[Rplots.pdf](https://github.com/ACCLAB/dabestr/files/3856883/Rplots.pdf)

**Comment by lczubiri** [comment]
Hi salve-amice,
Many thanks for your quick response.
I tried to follow your indications (except for the change in idx=list, for my real data I need that exact syntax, but thanks anyway!) but I must be doing something wrong, I get this messages:
![image](https://user-images.githubusercontent.com/57803688/69050191-c2f06180-0a01-11ea-9c78-2646276d18c2.png)
I'm mostly sure it has to do with how I changed the code, but don't know exactly why.

**Comment by lczubiri** [comment]
Hi salve-amice,
Many thanks for your quick response.
I tried what you suggested (except for the idx=list, because for my real data I'm comparing species 2-5 to species 1 within treatments, sorry for not giving further details, but thank you anyway) but I must be doing something wrong; I'm getting these messages:
![image](https://user-images.githubusercontent.com/57803688/69057441-2edac600-0a12-11ea-8964-b77e2d4cde7f.png)


But I've only edited the line you mentioned. Any hints for this? Thanks again.

**Comment by josesho** [comment]
Hi @lczubiri and @salve-amice ,

Thanks for being patient! This should be resolved with v0.2.5 (PR #77). Here's a quick recipe of how to use `colorRampPalette` along with `dabestr` to pass along a palette for > 12 colors.

```r
# Create example dataset.
set.seed(12345)
group.names <- paste("Group_", head(letters, 15), sep="")
my.data <- 
  tibble::tibble(
    group = rep(group.names, 10),
    value = rnorm(150, mean=10, sd=2.5)
  )

# Load the data into dabestr.
analyze.my.data <- dabest(my.data, group, value,
                          idx = list(c("Group_a", "Group_b", "Group_c"),
                                     c("Group_d", "Group_e", "Group_f"),
                                     c("Group_g", "Group_h", "Group_i"),
                                     c("Group_j", "Group_k", "Group_l"),
                                     c("Group_m", "Group_n", "Group_o")))



# All of the RColorBrewer palettes don't have more than 12 colors.
# Using `colorRampPalette()`, we can create a function that
# interpolates the needed colors for situations such as this.
# Here we use the 'Accent' palette found in RColorBrewer and which
# is accessible to ggplot2 via `scale_color_brewer()`
# See https://ggplot2.tidyverse.org/reference/scale_brewer.html#palettes
# for more info.

color.ramp.func <- grDevices::colorRampPalette(
                    RColorBrewer::brewer.pal(length(group.names), 'Accent')
                    )
my.colors <- color.ramp.func(length(group.names))
custom.pal <- setNames(my.colors, group.names)


# Plot the data.
plot(analyze.my.data, palette = custom.pal)
```

![image](https://user-images.githubusercontent.com/12838917/79844411-369dfc80-83ee-11ea-9514-9159a91d6be8.png)
